rebalance/rebalance.py

```diff
@@ -161,23 +165,64 @@ def rebalance(plugin, outgoing_scid, incoming_scid, msatoshi: Millisatoshi = Non
     start_ts = int(time.time())
     label = "Rebalance-" + str(uuid.uuid4())
     description = "%s to %s" % (outgoing_scid, incoming_scid)
     invoice = plugin.rpc.invoice(msatoshi, label, description, retry_for + 60)
     payment_hash = invoice['payment_hash']
-    success_msg = ""
-    try:
-        excludes = []
-        # excude all own channels to prevent unwanted shortcuts [out,mid,in]
-        mychannels = plugin.rpc.listchannels(source=my_node_id)['channels']
-        for channel in mychannels:
-            excludes += [channel['short_channel_id'] + '/0', channel['short_channel_id'] + '/1']
 
+    rpc_result = None
+    excludes = [my_node_id]  # excude all own channels to prevent shortcuts
+    nodes = {}               # here we store erring node counts
+    routes = {}              # trace all routes to suppress duplicates
+    maxhops = 1              # start with short routes and increase
+    msatfactor = plugin.msatfactor # start with high msatoshi factor to reduce
+                             # WIRE_TEMPORARY failures because of imbalances
+
+    # 'disable' maxhops filter if set to <= 0
+    if plugin.maxhops <= 0:
+        maxhops = 20
+
+    # trace stats
+    count = 0
+    count_sendpay = 0
+    time_getroute = 0
+    time_sendpay = 0
+
+    try:
         while int(time.time()) - start_ts < retry_for and not plugin.rebalance_stop:
-            r = plugin.rpc.getroute(incoming_node_id, msatoshi, riskfactor=1, cltv=9, fromid=outgoing_node_id, exclude=excludes)
+            count += 1
+            try:
+                time_start = time.time()
+                r = plugin.rpc.getroute(incoming_node_id,
+                                        fromid=outgoing_node_id,
+                                        exclude=excludes,
+                                        msatoshi=msatoshi * msatfactor,
+                                        maxhops=maxhops,
+                                        riskfactor=10, cltv=9)
+                time_getroute += time.time() - time_start
+            except RpcError as e:
+                # could not find route -> change params and restart loop
+                if e.method == "getroute" and e.error.get('code') == 205:
+                    # abort if we reached maxhop limit
+                    if plugin.maxhops > 0 and maxhops > plugin.maxhops:
+                        rpc_result = {'status':'error', 'message': 'No suitable routes found'}
+                        return cleanup(plugin, label, payload, rpc_result)
+                    # reduce msatfactor to look for smaller channels now
+                    msatfactor -= 1
+                    if msatfactor < 1:
+                        # when we reached neutral msat factor:
+                        # increase maxhops and restart with msatfactor
+                        maxhops = maxhops + 1
+                        msatfactor = plugin.msatfactor
+                    continue
+                else:
+                    raise e
```

Comment:
Almost perfect. First increase `maxhops`, then check if `maxhops` is too big! Otherwise it will run once with `maxhops = plugin.maxhops + 1`.
```suggestion
            except RpcError as e:
                # could not find route -> change params and restart loop
                if e.method == "getroute" and e.error.get('code') == 205:
                    # reduce msatfactor to look for smaller channels now
                    msatfactor -= 1
                    if msatfactor < 1:
                        # when we reached neutral msat factor:
                        # increase maxhops and restart with msatfactor
                        maxhops = maxhops + 1
                        msatfactor = plugin.msatfactor
                    # abort if we reached maxhop limit
                    if plugin.maxhops > 0 and maxhops > plugin.maxhops:
                        rpc_result = {'status':'error', 'message': 'No suitable routes found'}
                        return cleanup(plugin, label, payload, rpc_result)
                    continue
                else:
                    raise e
```

Reply:
Okay, I switched the blocks back again. now just without the 'continue' Guess your right. thanks